Revise details for hostPath volumes (#42510)
kubernetes/website: content/en/docs/concepts/storage/volumes.md

or as read-write, because:

* Access to the host filesystem can expose privileged system credentials (such as for the kubelet) or privileged APIs
  (such as the container runtime socket), that can be used for container escape or to attack other
  parts of the cluster.
* Pods with identical configuration (such as created from a PodTemplate) may
  behave differently on different nodes due to different files on the nodes.
{{< /warning >}}

Some uses for a `hostPath` are:

* running a container that needs access to node-level system components
  (such as a container that transfers system logs to a central location,
  accessing those logs using a read-only mount of `/var/log`)
* making a configuration file stored on the host system available read-only
  to a {{< glossary_tooltip text="static pod" term_id="static-pod" >}};
  unlike normal Pods, static Pods cannot access ConfigMaps

#### `hostPath` volume types

In addition to the required `path` property, you can optionally specify a
`type` for a `hostPath` volume.

The available values for `type` are:

<!-- empty string represented using U+200C ZERO WIDTH NON-JOINER -->

| Value | Behavior |
|:------|:---------|

> | `‌` | Empty string (default) is for backward compatibility, which means that no checks will be performed before mounting the `hostPath` volume. |

Reviewer: Maybe `""` is more intuitive for an empty string?

> | `‌` | Empty string (default) is for backward compatibility, which means that no checks will be performed before mounting the `hostPath` volume. |

Author: | `‌""` | Empty string (default) is for backward compatibility, which means that no checks will be performed before mounting the `hostPath` volume. |

### hostPath {#hostpath}

A `hostPath` volume mounts a file or directory from the host node's filesystem
into your Pod. This is not something that most Pods will need, but it offers a
powerful escape hatch for some applications.

> {{< warning >}}

Reviewer 2: One more warning potentially worth adding, based on a few production bugs I've encountered:

Hostpath on "/" is strongly discouraged. Not only does it present security risks, but it can interfere with Kubernetes volume unmounting when used with the default "None" [mount propagation](https://kubernetes.io/docs/concepts/storage/volumes/#mount-propagation) . Use "HostToContainer" mount propagation if you must.

Author: Can we save that for a different PR? This PR has a different aim.

Reviewer: I tend to agree with @msau42 that adding that note (warning) would be desirable.

Author: I agree! Can we save that change for a different PR? If not: why not?

Using the `hostPath` volume type presents many security risks.
If you can avoid using a `hostPath` volume, you should. For example,
define a [`local` PersistentVolume](#local), and use that instead.

If you are restricting access to specific directories on the node using
admission-time validation, that restriction is only effective when you
additionally require that any mounts of that `hostPath` volume are
**read only**. If you allow a read-write mount of any host path by an
untrusted Pod, the containers in that Pod may be able to subvert the
read-write host mount.

---

Take care when using `hostPath` volumes, whether these are mounted as read-only
or as read-write, because:

> * Access to the host filesystem can expose privileged system credentials (such as for the Kubelet) or privileged APIs

Reviewer: We don't capitalize the `K` in kubelet.

> * Access to the host filesystem can expose privileged system credentials (such as for the Kubelet) or privileged APIs

Author: ```suggestion
* Access to the host filesystem can expose privileged system credentials (such as for the kubelet) or privileged APIs
```

  (such as the container runtime socket), that can be used for container escape or to attack other
  parts of the cluster.

> * Pods with identical configuration (such as created from a PodTemplate) may

Reviewer: ```suggestion
* Pods with identical configuration such as created from the same PodTemplate may
```

Author: The suggestion here doesn't read like idiomatic English. Can you clarify what's not right with what I proposed @tengqm?

Reviewer: Answered in another comment. I think we may want to avoid parenthesis when possible. They interrupt the sentences and make them don't flow well. It can be used occasionally for reminding something, but having so many parens in these bullets could be a problem. Never read docs like this before.

Author: This is moved text; I've left it as-is but changed where it appears in the page.

>   behave differently on different nodes due to different files on the nodes

Reviewer 3: nit: missing period at end of sentence

Reviewer 3: Since these items are important, should the content appear in a separate page (so that the content is not lost) or provided as a list of known restrictions/limitations?

{{< /warning >}}

For example, some uses for a `hostPath` are:

* running a container that needs access to node-level system components

>   (such as reading system logs using a read-only mount of `/var/log`)

Reviewer: ```suggestion
  such as reading system logs using a read-only mount of `/var/log`
```

Author: Making the change proposed here could make readers backtrack. I'll reword, but the parens are needed.

Reviewer: I cannot recall where I read it ... but the general guideline for technical writing is to avoid parenthesis when possible. We lost nothing by removing them.

Author: We do lose something. It otherwise becomes a [garden path sentence](https://en.wikipedia.org/wiki/Garden-path_sentence) where someone reading in English initially thinks that “reading system logs” qualifies which kind of node-level system components the container needs access to.

I'll take a rewritten phrase, sure, but this suggestion introduces something a bit like a code regression.

| `File` | A file must exist at the given path |
| `Socket` | A UNIX socket must exist at the given path |
| `CharDevice` | _(Linux nodes only)_ A character device must exist at the given path |
| `BlockDevice` | _(Linux nodes only)_ A block device must exist at the given path |

{{< caution >}}
The `FileOrCreate` mode does **not** create the parent directory of the file. If the parent directory
of the mounted file does not exist, the pod fails to start. To ensure that this mode works,
you can try to mount directories and files separately, as shown in the
[`FileOrCreate` example](#hostpath-fileorcreate-example) for `hostPath`.
{{< /caution >}}

Some files or directories created on the underlying hosts might only be
accessible by root. You then either need to run your process as root in a
[privileged container](/docs/tasks/configure-pod-container/security-context/)
or modify the file permissions on the host to be able to read from
(or write to) a `hostPath` volume.

#### hostPath configuration example

{{< tabs name="hostpath_examples" >}}

> {{< tab name="Linux node" codelang="yaml" >}}

Reviewer: ```suggestion
{{< tab name="Linux node" >}}
```

Author: Why not mark it as YAML?

Reviewer: Please read the comment https://github.com/kubernetes/website/pull/42510#discussion_r1343830111.

Marking the whole tab as YAML doesn't buy us anything. At the same time, it create a huge problem when there are comments in the YAML file.

Author: Ah, OK. That makes sense with context. I've revised the page to put that explanatory text outside the example manifest.

> ---
> # This manifest mounts /data/foo on the host as /foo inside the
> # single container that runs within the hostpath-example-linux Pod.
> #
> # The mount into the container is read-only.

Reviewer: ````suggestion

The manifest mounts `/data/foo` on the host as `/foo` inside the
single container that runs within the hostpath-example-linux Pod.
The mount into the container is read-only.

```yaml
---
````

Author: Why's this better? I can't guess.

Reviewer: It is "better" because it would help the downstream localization. Lemme explain it in details.

Suppose we have a code snippet (in this case, the whole tab is a snippet) like this:

````
```yaml
# some comments
the:
  real:
    stuff
```
````

What we the localization team is expected to translate the comment line. We do translations by commenting out the original English text (using HTML comment syntax that is acceptable to Markdown as well) and then adding translated texts. Comments in code snippets are difficult to handle, since they are supposed to be rendered in a verbatim way. For example we cannot do this:

````
```yaml
<!-- some comments -->
# 一些注釋
the:
  real:
    stuff
```
````

We had to comment out the whole snippet out and add a new one, even if there is only one line of English comment there. That is:

````
<!--
```yaml
# some comments
the:
  real:
    stuff
```
-->
```yaml
# 一些注釋
the:
  real:
    stuff
```
````

This is kind of annoying. If those inline comments are extracted into standalone paragraphs, the localization becomes very easy. We simply don't need to touch the code snippet. We can focus on the texts instead. I mean the following case:

````
General introductions to the following YAML goes here, then the example:
```yaml
the:
  yaml:
    or: whatever code
```
````

can be handled cleanly as:

````
<!--
General introductions to the following YAML goes here, then the example:
-->
這裡是對下面的 YAML 的一般性介紹，不要寫成行內注釋：

```yaml
the:
  yaml:
    or: whatever code
```
````

Hopefully this explains why I am kinda picky on inline comments unless necessary.

Author: I think the snippet works better how I've proposed it.

If the overall localizations subproject recommends a particular approach then fine, let's do it.
If just the Chinese localization team prefer it - I think that isn't strong enough grounds to revise this PR. I'm wary of informally adopting style guide changes that suit single localizations. Formal adoption with discussion _is_ fine.

Reviewer: @sftim Are you willing to revise the structure as suggested?
The suggested layout doesn't hurt readability, but it at least makes the downstream localization easier.

Author: OK, I'll have a go.

Author: How's the revised PR?

apiVersion: v1
kind: Pod
metadata:
  name: hostpath-example-linux
spec:
  os: { name: linux }
  nodeSelector:
    kubernetes.io/os: linux
  containers:
  - name: example-container
    image: registry.k8s.io/test-webserver
    volumeMounts:
    - mountPath: /foo
      name: example-volume
      readOnly: true
  volumes:
  - name: example-volume
    # mount /data/foo, but only if that directory already exists
    hostPath:
      path: /data/foo # directory location on host
      type: Directory # this field is optional
{{< /tab >}}

> {{< tab name="Windows node" codelang="yaml" >}}
> ---
> # This manifest mounts C:\Data\foo on the host as C:\foo, inside the
> # single container that runs within the hostpath-example-windows Pod.
> #
> # The mount into the container is read-only.

Reviewer: ````suggestion
{{< tab name="Windows node" >}}

This manifest mounts `C:\Data\foo` on the host as `C:\foo`, inside the
single container that runs within the hostpath-example-windows Pod.
The mount into the container is read-only.

```yaml
---
````

Author: How's the revised PR?

apiVersion: v1
kind: Pod
metadata:
  name: hostpath-example-windows
spec:
  os: { name: windows }
  nodeSelector:
    kubernetes.io/os: windows
  containers:
  - name: example-container
    image: microsoft/windowsservercore:1709
    volumeMounts:
    - name: example-volume
      mountPath: "C:\\foo"
      readOnly: true
  volumes:
  - name: example-volume
    hostPath:
      path: "C:\\Data\\foo"
{{< /tab >}}
{{< /tabs >}}

#### hostPath FileOrCreate configuration example {#hostpath-fileorcreate-example}

> This manifest defines a Pod that mounts `/var/local/aaa`

Reviewer: ```suggestion
The following manifest defines a Pod that mounts `/var/local/aaa`
```

inside the single container in the Pod. If the node does not
already have a path `/var/local/aaa`, the kubelet creates
it as a directory and then mounts it into the Pod.

If `/var/local/aaa` already exists but is not a directory,
the Pod fails. Additionally, the kubelet attempts to make
a file named `/var/local/aaa/1.txt` inside that directory
(as seen from the host); if something already exists at
that path and isn't a regular file, the Pod fails.

> 
> Here's the example manifest:

Reviewer: ```suggestion
```

Author: Please explain. I'm an approver, and _I_ can't guess why you want this change made.
